src/llvm_ir_generator.cpp

```diff
@@ -163,39 +163,83 @@ void LLVMIRGenerator::Visit(const DeclStmtNode& decl_stmt) {
-  auto addr = builder_.CreateAlloca(var_type);
-  if (decl.init) {
-    decl.init->Accept(*this);
-    auto val = val_recorder.ValOfPrevExpr();
-    builder_.CreateStore(val, addr);
+  if (decl.is_global) {
+    auto global = module_.getOrInsertGlobal(decl.id, var_type);
+    auto global_var = llvm::dyn_cast<llvm::GlobalVariable>(global);
+    llvm::Constant* const_val = nullptr;
+    if (decl.init) {
+      decl.init->Accept(*this);
+      auto val = val_recorder.ValOfPrevExpr();
+      const_val = llvm::dyn_cast<llvm::Constant>(val);
+      assert(const_val);
+    } else {
+      const_val = llvm::ConstantInt::get(builder_.getInt32Ty(), 0, true);
+    }
+    global_var->setInitializer(const_val);
+    id_to_val[decl.id] = global_var;
+  } else {
+    auto addr = builder_.CreateAlloca(var_type);
+    if (decl.init) {
+      decl.init->Accept(*this);
+      auto val = val_recorder.ValOfPrevExpr();
+      builder_.CreateStore(val, addr);
+    }
+    id_to_val[decl.id] = addr;
   }
-  id_to_val[decl.id] = addr;
 }
 
 void LLVMIRGenerator::Visit(const ArrDeclNode& arr_decl) {
-  auto arr_type = builder_helper_.GetLLVMType(*(arr_decl.type));
-  auto base_addr = builder_.CreateAlloca(arr_type);
-  id_to_val[arr_decl.id] = base_addr;
+  auto type = builder_helper_.GetLLVMType(*(arr_decl.type));
+  if (arr_decl.is_global) {
+    auto global_arr = module_.getOrInsertGlobal(arr_decl.id, type);
+    id_to_val[arr_decl.id] = global_arr;
+  } else {
+    auto addr = builder_.CreateAlloca(type);
+    id_to_val[arr_decl.id] = addr;
+  }
 
   auto arr_decl_type = dynamic_cast<ArrType*>(arr_decl.type.get());
-  for (auto i = std::size_t{0}, e = arr_decl_type->len(); i < e; ++i) {
-    if (i < arr_decl.init_list.size()) {
+  // This vector stores the initialize values for an array.
+  std::vector<llvm::Constant*> arr_elems{};
+  for (auto i = std::size_t{0}, e = arr_decl_type->len(),
+            init_len = arr_decl.init_list.size();
+       i < e; ++i) {
+    if (i < init_len) {
       auto& arr_init = arr_decl.init_list.at(i);
       arr_init->Accept(*this);
     }
 
-    auto res_addr =
-        builder_.CreateConstInBoundsGEP2_32(arr_type, base_addr, 0, i);
-
-    if (i < arr_decl.init_list.size()) {
+    if (i < init_len) {
       auto init_val = val_recorder.ValOfPrevExpr();
-      builder_.CreateStore(init_val, res_addr);
+      if (arr_decl.is_global) {
+        auto const_val = llvm::dyn_cast<llvm::Constant>(init_val);
+        arr_elems.push_back(const_val);
+      } else {
+        auto res_addr = builder_.CreateConstInBoundsGEP2_32(
+            type, id_to_val.at(arr_decl.id), 0, i);
+        builder_.CreateStore(init_val, res_addr);
+      }
     } else {
       // set remaining elements as 0
       auto zero = llvm::ConstantInt::get(builder_.getInt32Ty(), 0, true);
-      builder_.CreateStore(zero, res_addr);
+      if (arr_decl.is_global) {
+        arr_elems.push_back(zero);
+      } else if (!arr_decl.is_global && 0 < init_len) {
```

Comment:
Since `init_len` is unsigned, use `!= 0` to better convey the condition that it should not be 0.
```suggestion
      } else if (!arr_decl.is_global && init_len != 0) {
```